src/librbd/io/ImageRequest.cc

```diff
@@ -543,6 +542,59 @@ void ImageDiscardRequest<I>::update_stats(size_t length) {
   image_ctx.perfcounter->inc(l_librbd_discard_bytes, length);
 }
 
+template <typename I>
+int ImageDiscardRequest<I>::prune_object_extents(
+    ObjectExtents* object_extents) const {
+  if (m_discard_granularity_bytes == 0) {
+    return 0;
+  }
+
+  // Align the range to discard_granularity_bytes boundary and skip
+  // and discards that are too small to free up any space.
+  //
+  // discard_granularity_bytes >= object_size && tail truncation
+  // is a special case for filestore
+  bool prune_required = false;
+  auto object_size = this->m_image_ctx.layout.object_size;
+  auto discard_granularity_bytes = std::min<uint64_t>(
+    m_discard_granularity_bytes, object_size);
```

Comment:
Can this become a simple std::min() now?

Reply:
I need to force the variable to 64 bits so that ```p2roundup``` works correctly (it fails spectacularly if you put in different data sizes since a negative 64-bit number bitwise and-ed with a negative 32-bit number is still a huge negative 64-bit number).

Reply:
Right, uint32_t won't be sign-extended after negation...

I can send a PR to fix p2*() signatures, but in the meantime I think this definitely deserves a comment -- "force uint64_t for p2*()" or something like that.

Reply:
I actually added [1] to ensure the sizes were equal to avoid the issue

[1] https://github.com/ceph/ceph/pull/26432/files#diff-a4882de2d280b5a2e46dfc9e6ea5e90cR570

Reply:
Makes sense, I'll change these calls to "p2*<uint64_t>(...)" and remove it in my PR.